File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/datastructures/Problem9012.java
package com.keunwon.algorithm.datastructures;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Problem9012 {

    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        while (N -- > 0) {
            Stack<Character> stack = new Stack<>();
            String input = br.readLine();
            boolean isSuccess = true;

            for (int i = 0; i < input.length(); i++) {
                char c = input.charAt(i);

                if (c == '(') {
                    stack.add(c);
                    continue;
                }

                if (stack.isEmpty()) {
                    isSuccess = false;
                    break;
                }

                stack.pop();
            }

            if (stack.size() != 0) {
                isSuccess = false;
            }

            System.out.println(isSuccess ? "YES" : "NO");
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem9012.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 괄호
 * Level: 실버-4
 **/
class Problem9012 {
    fun solution(word: String): String {
        val stack = Stack<Char>()

        for (c in word) {
            if (c == '(') stack.push(c)
            else if (c == ')') {
                if (stack.isNotEmpty()) stack.pop()
                else return "NO"
            }
        }
        return if (stack.isEmpty()) "YES" else "NO"
    }
}

fun main() {
    val t = readLine()!!.toInt()
    repeat(t) {
        val word = readLine()!!
        Problem9012().solution(word).also { println(it) }
    }
}
